### ISSUE TITLE
RFC: Inject slab allocators into OSTD

### ISSUE THREAD

**Issue by tatetian** [asterinas/asterinas#1555]
### Summary

This RFC introduces to OSTD new abstractions for _slabs_ so that efficient _slab caches_ can be implemented in safe Rust by OSTD users. These slab caches can be used to implement an efficient slab-based heap allocator, which, in turn, can be _injected_ into OSTD as Rust's global heap allocator. This approach enables OSTD-based kernels to enjoy efficient, slab-based heap allocators, while keeping most complexities out of the OSTD.

### Background and Motivation

Implementing an efficient heap allocator is no easy task. TCMalloc, Google's custom implementation of C's `malloc`, has a codebase of 60K lines of code. In the context of OS kernels, a widely accepted object allocation technique is _slab allocation_. In Linux, there are historically three slab allocators: SLAB, SLOT, and SLUB; the file `slub.c` alone has over 7K lines of code.

In Asterinas, the current heap allocator is quite naive, less than 1K lines of code. It lacks of some key elements that are commonly adopted by UNIX-style kernels like Linux:
1. Slabs. A slab is one or multiple contiguous pages that are divided into a number of fixed-size slots, each of which can be used to store objects of a certain type or size.
2. Caches. A slab cache manages a set of slabs that can be full, partial, or empty. Keeping more slabs benefits performance but leads to more memory consumption. So good (and probably complex) strategies may be employed to balance the goals of high performance and low memory consumption.
3. Per-CPU data structures. On a SMP machine, a key performance consideration is to use per-CPU data structures to avoid lock contention across different threads.

Following the framekernel's philosophy of keeping OSTD as small as possible, we want to keep the most complexity introduced by a SMP-aware, highly-efficient slab allocator out of OSTD. To the best of our knowledge, no prior OS kernels are designed to partition the implementation of a slab allocator into safety-relevant and -irrelevant parts. So this RFC are up to this task.

### Proposal

In this RFC, we claim that adding abstractions for slabs and slab slots to OSTD is sufficient to enable developing efficient slab allocators in safe Rust. And this approach won't introduce any soundness problem.

![image](https://github.com/user-attachments/assets/634cfc9d-a86f-4e8d-a365-ea3f93096ffd)

#### Core abstractions

The core abstraction is `Slab`.

```rust
//! A slab is one or multiple contiguous pages
//! that are divided into a number of fixed-sized slots,
//! each of which can be used as the storage for an object
//! whose size is no greater than the slot size.
pub struct Slab<const SLOT_SIZE: usize, Ext> {
    page: NonNull<u8>,
    phantom: PhantomData<Ext>
}
```

`Slab` is allocated by OSTD, just like other types of pages.

```rust
impl<const SLOT_SIZE: usize, Ext> Slab<SLOT_SIZE, Ext> {
    /// Allocate a page-sized slab with the user-given slab extension.
    pub fn alloc(
        slot_recyle_fn: SlotRecycleFn<SLOT_SIZE>,
        extension: Ext
    ) -> Option<Self> {
        todo!("
            Step 1. Allocate a new page for slab with the specified metadata and extension.
            Step 2. Partition the slab as an array of FreeSlabSlot.
            Step 3. Link all FreeSlabSlots into a list.
        ")
    }
}
```

Let's ignore the two arguments of `Slab::alloc` for now. We will come back to them later.

With `Slab`, we can allocate and deallocate slab slots.

```rust
impl<const SLOT_SIZE: usize, Ext> Slab<SLOT_SIZE, Ext> {
    /// Allocates a free slot.
    pub fn new_slot(&mut self) -> Option<FreeSlabSlot<SLOT_SIZE>> { todo!() }

    /// Recycles a free slot.
    pub fn recycle_slot(&mut self, free_slot: FreeSlabSlot<SLOT_SIZE>) { todo!() }
}
```

A `FreeSlabSlot` can be converted into `Box<T>` or `Arc<T>`.

```rust
/// A free slab slot.
pub struct FreeSlabSlot<const SLOT_SIZE: usize> {
    ptr: NonNull<[u8; SLOT_SIZE]>,
}

impl<const SLOT_SIZE: usize> FreeSlabSlot<SLOT_SIZE> {
    pub fn into_box<T>(self, obj: T) -> Box<T> { todo!() }
    pub fn from_box<T>(boxed_obj: Box<T>) -> Self { todo!() }

    pub fn into_arc<T>(self, data: T) -> Arc<T> { todo!() }
    pub fn from_arc<T>(arc: Arc<T>) -> Self { todo!() }
}
```

OSTD ensures that the memory behind a `FreeSlabSlot<SLOT_SIZE>` is free for use and satisfies the size and alignment requirements by `Box<T>` or `Arc<T>`. Furthermore, OSTD ensures that the memory behind a `Slab` is valid as long as the `Slab` is alive or any `FreeSlabSlot`s allocated from it are alive.

#### Slab metadata and extensions

In our descriptions above, we have intentionally skip the arguments of `Slab::alloc`. What are the purposes of `slot_recyle_fn` and `extension`? Where are they stored?

```rust
impl<const SLOT_SIZE: usize, Ext> Slab<SLOT_SIZE, Ext> {
    /// Allocate a page-sized slab with the user-given slab extension.
    pub fn alloc(
        slot_recyle_fn: SlotRecycleFn<SLOT_SIZE>,
        extension: Ext
    ) -> Option<Self> { todo!() }
}
```

The answer is the slab metadata, `SlabMeta`.

```rust
/// The metadata for a slab.
#[repr(C)]
pub(crate) struct SlabMeta<const SLOT_SIZE: usize, Ext> {
    // Keeps track of the available `FreeSlabSlot` that can be allocated
    free_list: AtomicPtr<FreeSlabSlot>,
    // Counts the number of `FreeSlabSlot` that have been allocated.
    //
    // The parent slab cannot be dropped unless this count is zero.
    nr_inuse_slots: AtomicU16,
    // A recycle callback function that will be invoked by the heap allocator.
    //
    // Different "heap allocations" may come from different slabs or slab caches.
    // This callback 
    slot_recyle_fn: SlotRecycleFn,
    // The extension provided by the OSTD user is stored in two fields.
    //
    // An implementation of slab allocator can customize this extension field
    // to associate some extra data with each instance of Slab.
    //
    // The first field stores the vtable of `Ext`` as an `dyn Any` trait object.
    extension_vtable: DynMetadata<dyn Any + 'static>,
    // The second first stores the real content of `Ext`.
    extension: Ext,
}

/// The callback function to recycle a free slot back to its owner
/// (e.g., a slab cache).
pub type SlotRecycleFn<const SLOT_SIZE: usize> = fn(
    /* slot: */FreeSlabSlot<SLOT_SIZE>,
    /* extension: */&dyn Any,
    /* pin_cpu_guard: */&dyn PinCurrentCpu,
);
```

The type definition is extensively documented. I hope they are clear enough.

Each `SlabMeta` is associated with a `Slab`, which corresponds to one or multiple contiguous pages. Thus, `SlabMeta` can be stored using [the page meta mechanism](https://github.com/asterinas/asterinas/blob/main/ostd/src/mm/page/meta.rs), without heap allocations.

#### Using slab allocators for heap

OSTD users are free to implement arbitrary slab allocators in safe Rust by leveraging the two abstractions of `Slab` and `FreeSlabSlot`. If these slab allocators implement the `SlabSlotAlloc` trait, then they can be injected into OSTD as the heap allocator.

```rust
/// Injects an array of slab allocators of different slot sizes.
pub fn inject_slab_allocators(slab_alloc_array: SlabAllocators) {
    self.HEAP_ALLOC.inject_slab_allocators(slab_allocators)
}

/// An array of slab allocators.
pub struct SlabAllocators {
    pub size16: &'static dyn SlabSlotAlloc<16>,
    pub size32: &'static dyn SlabSlotAlloc<32>,
    // ...
    pub size2048: &'static dyn SlabSlotAlloc<2048>,
}

/// Represents any slab allocator.
pub trait SlabSlotAlloc<const OBJ_SIZE: usize> {
    fn alloc(&self, current_cpu: &dyn PinCurrentCpu) -> Option<FreeSlabSlot<OBJ_SIZE>>;
}
```

Although a heap allocator based on these injected slab caches cannot determine which memory regions to use, the type `FreeSlabSlot` promises that the memory behind is valid for heap usage.

#### The early heap

Heap allocations are necessary during the bootstrap and initialization process of OSTD, which happens before slab allocator injection. So how can these early heap allocations be satisfied? The answer is the early heap.
```rust
mod early_heap;

use early_heap::EarlyHeapAlloc;

#[global_allocator]
static HEAP_ALLOC: HeapAlloc = {
    // SAFETY: The global heap allocator is created only once.
    unsafe {
        HeapAlloc::new()
    }
};

struct HeapAlloc {
    have_injected_slabs: AtomicBool,
    backend: HeapAllocBackend,
}

struct HeapAllocBackend {
    early_heap: SpinLock<EarlyHeapAlloc>,
    slab_caches: Once<SlabAllocators>,
}

unsafe impl GlobalAlloc for HeapAlloc {
    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
        if layout.size >= MAX_SLAB_SLOT {
            return todo!("use the page allocator directly, instead of slab allocators");
        }

        let slot_size = determine_slot_size(layout.size());

        let slab_allocators = match self.current_backend() {
            EarlyHeap(early_heap) => {
                let mut early_heap_guard = early_heap.lock();
                return early_heap.alloc(slot_size);
            }
            SlabCaches(slab_allocators) => slab_allocators,
        };

        todo!("use the slab allocators")
    }

    unsafe fn dealloc(&self, slot_ptr: *mut u8, layout: Layout) {
        if layout.size >= MAX_SLAB_SLOT {
            return todo!("use the page allocator directly, instead of slab allocators");
        }

        let slot_size = self.determine_slot_size(layout.size());

        let slab_allocators = match self.current_backend() {
            EarlyHeap(early_heap) => {
                let mut early_heap_guard = early_heap.lock();
                // SAFETY: the memory represented by pointer and size 
                // is valid and must have been allocated from the early heap.
                unsafe {
                    early_heap.dealloc(slot_ptr, slot_size)
                };
            }
            SlabCaches(slab_allocators) => slab_allocators,
        };

        // We MUST NOT use the injected slab caches
        // to deallocate memory allocated from the early heap.
        if early_heap::contains_ptr(slot_ptr) {
            // It is ok to simply "leak" memory in the early heap,
            // instead of deallocating it from the early heap.
            return;
        }

        todo!("use the slab allocators")
    }
}
```

A draft implementation of the `Slab`, `FreeSlabSlot`, and the heap allocation can be found [here](https://github.com/tatetian/injected_slab_allocators/tree/main/src/ostd/heap).

### Example implementations of slab allocators

In this part, we will give three implementation of slab allocators, with increasingly higher complexities and performance. 

#### Version 0: A single-page slab allocator

https://github.com/tatetian/injected_slab_allocators/blob/ae89d1b848724da95f1456a0dc77fe57bc1e610e/src/kernel/slab_v1.rs

#### Version 1: A scalable, single-page slab allocator

https://github.com/tatetian/injected_slab_allocators/blob/ae89d1b848724da95f1456a0dc77fe57bc1e610e/src/kernel/slab_v2.rs

#### Version 2: A lockless, scalable, single-page slab allocator

https://github.com/tatetian/injected_slab_allocators/blob/ae89d1b848724da95f1456a0dc77fe57bc1e610e/src/kernel/slab_v3.rs

**Comment by junyang-zh** [comment]
```rust
impl<const SLOT_SIZE: usize> FreeSlabSlot<SLOT_SIZE> {
    pub fn into_box<T>(self, obj: T) -> Box<T> { todo!() }
    pub fn from_box<T>(boxed_obj: Box<T>) -> Self { todo!() }

    pub fn into_arc<T>(self, data: T) -> Arc<T> { todo!() }
    pub fn from_arc<T>(arc: Arc<T>) -> Self { todo!() }
}
```

These 4 APIs are impossible to be implemented, see [doc](https://doc.rust-lang.org/beta/alloc/boxed/struct.Box.html).

I guess that you want `pub fn into_box<T>(self, obj: T) -> Box<T, A>` where `A` is slab. Unfortunately A must be a singleton type, and `Slab` is not a singleton type.

So Rust can't tell which function should we call when a box goes out of scope. Even worse Rust will hand the pointer to the global allocator for recycling the box. Then you'll probably never be able to implement the slab cache.

**Comment by tatetian** [comment]
> ```rust
> impl<const SLOT_SIZE: usize> FreeSlabSlot<SLOT_SIZE> {
>     pub fn into_box<T>(self, obj: T) -> Box<T> { todo!() }
>     pub fn from_box<T>(boxed_obj: Box<T>) -> Self { todo!() }
> 
>     pub fn into_arc<T>(self, data: T) -> Arc<T> { todo!() }
>     pub fn from_arc<T>(arc: Arc<T>) -> Self { todo!() }
> }
> ```
> 
> These 4 APIs are impossible to be implemented, see [doc](https://doc.rust-lang.org/beta/alloc/boxed/struct.Box.html).
> 
> I guess that you want `pub fn into_box<T>(self, obj: T) -> Box<T, A>` where `A` is slab.

No, the proposed API has been defined accurately: it is meant to return `Box<T>` and `Arc<T>` without the custom allocator type.

> Unfortunately A must be a singleton type, and `Slab` is not a singleton type.

This is the tricky (or brilliant) part of the proposed design. It allows OSTD users to instantiate arbitrary number of slab allocators of arbitrary implementations without "polluting" the type of `Box<T>` or `Arc<T>`. This is possible because **although slab allocators are not singleton, the slab mechanism can be viewed as a singleton.** The `Slab` type provided by OSTD manages the life cycle of slabs and their slab slots.

> So Rust can't tell which function should we call when a box goes out of scope. Even worse Rust will hand the pointer to the global allocator for recycling the box. Then you'll probably never be able to implement the slab cache.

Check out [the draft code for the new global heap allocator](https://github.com/tatetian/injected_slab_allocators/blob/ae89d1b848724da95f1456a0dc77fe57bc1e610e/src/ostd/heap/mod.rs#L135). All heap allocations, including all `Arc<T>` and `Box<T>`, are returned to the global heap allocator, `HeapAlloc`, which converts a given `ptr` to the corresponding `FreeSlabSlot`. Each `FreeSlabSlot` is associated with a `Slab` and its `SlabMeta`. The latter one allows the global heap allocator to retrieve a per-slab allocator callback, which can then be used to return the `FreeSlabSlot` to the corresponding slab allocator. Depending on the implementation, the slab allocator probably may choose to do one of the two actions:
1. Link the given `FreeSlabSlot` to a free list so that it can be reused cheaply;
2. Find out the owner slab of the slot and then call `Slab::recycle_slot`.
